lib/spack/spack/package_base.py

```diff
@@ -1462,6 +1462,14 @@ def do_patch(self):
             tty.msg("No patches needed for {0}".format(self.name))
             return
 
+        # Make sure staged files are writable to avoid issues with file permissions
+        # on Lustre
+        for dir_path, _, sub_files in os.walk(self.stage.source_path):
+            for sub_file in sub_files:
+                sub_path = os.path.join(dir_path, sub_file)
+                perm = os.stat(sub_path).st_mode
+                os.chmod(sub_path, perm | 0o200)
```

Comment:
I'd like to see this done in `PackageBase.do_stage` or even as part of the method returned by `compression.decompressor_for`. I thought we set sticky bits (`S_ISVTX`) as part of `Stage.create` (we do at least as part of our post-install hooks modifying the installation prefix - but maybe that wouldn't apply anyhow on lustre). We at least make sure the stage directory we create is writable, so it would make sense IMO to extend that to the files resulting from `Stage.expand_archive`. Some care might be needed to avoid repeated `chmod`ing of files after each resource is relocated into the root stage dir (in the case that the package has stages).

The other thing I think is worth mentioning is the `config:build_stage` option, which you can use to set where the stage root is (e.g. on another filesystem). By default, we should be trying to build in a local fast space like `/tmp`, so I'm curious if you explicitly set this to lustre (or if you happen to know that `/tmp` is unavailable). To be clear, it is perfectly reasonable to use lustre for staging (this can be useful for example if you want to examine the logs of failed builds after dispatching the build within an allocation) and we should support it (I thought mentioning this option might be useful though).

Is that enough to go off of?